covid_moonshot_ml/data/dataset.py

```diff
@@ -21,40 +29,236 @@ def __init__(self, str_fns, compounds, ignore_h=True, lig_resn="LIG"):
         lig_resn : str, default='LIG'
             The residue name for the ligand molecules in the PDB files. Used to
             identify which atoms belong to the ligand
+        extra_dict : dict[str, dict], optional
+            Extra information to add to each structure. Keys should be
+            compounds, and dicts can be anything as long as they don't have the
+            keys ["z", "pos", "lig", "compound"]
+        num_workers : int, default=1
+            Number of cores to use to load structures
         """
+        super(DockedDataset, self).__init__()
+
+        ## Function to extract from extra_dict (just to make the list
+        ##  comprehension look a bit nicer)
+        get_extra = lambda compound: (
+            extra_dict[compound]
+            if (extra_dict and (compound in extra_dict))
+            else None
+        )
+        mp_args = [
+            (fn, compound, ignore_h, lig_resn, get_extra(compound))
+            for i, (fn, compound) in enumerate(zip(str_fns, compounds))
+        ]
+
+        if num_workers > 1:
+            import multiprocessing as mp
+
+            n_procs = min(num_workers, mp.cpu_count(), len(mp_args))
+            with mp.Pool(n_procs) as pool:
+                all_structures = pool.starmap(
+                    DockedDataset._load_structure, mp_args
+                )
+        else:
+            all_structures = [
+                DockedDataset._load_structure(*args) for args in mp_args
+            ]
+
+        self.compounds = {}
+        self.structures = []
+        for i, (compound, struct) in enumerate(zip(compounds, all_structures)):
+            try:
+                self.compounds[compound].append(i)
+            except KeyError:
+                self.compounds[compound] = [i]
+            self.structures.append(struct)
+
+    @staticmethod
+    def _load_structure(fn, compound, ignore_h, lig_resn, extra_dict):
```

Comment:
Insert docstring for this fn

Reply:
added in [51b4b38](https://github.com/choderalab/covid-moonshot-ml/pull/81/commits/51b4b386c8df7364c871a0c7e73e602f94bda601)